index.js

```diff
@@ -1,186 +1,259 @@
+        });
+
+      }, function(error) {
+        reject(error);
+      });
 
-    process.once('uncaughtException', function(error) {
-      if (process.listeners('uncaughtException').length === 0) {
-        simProcess.kill('SIGTERM');
-        throw error;
+  });
+}
+
+// Find a simulator that matches the options
+function findSimulator(options) {
+
+  return new Promise(function(resolve, reject) {
+
+    findSimulators(options).then(function(results) {
+
+      if (!results || results.length === 0) {
+        reject(new Error('No simulators installed, or cannot find them'));
       }
+
+      // just returning the first result for now
+      resolve(results[0]);
+
+    }, function(error) {
+      reject(error);
     });
-  }
 
-  if (opts.exit) {
-    simProcess.unref();
-  }
-  defer.resolve(simProcess);
-  return defer.promise;
+  });
+
 }
 
-function createClient(simulator) {
-  var deferred = Q.defer();
-  var client = new FirefoxClient();
-  client.connect(simulator.port, function(err) {
-    if (err) {
-      deferred.reject(err);
+
+function findAvailablePort(preferredPort) {
+
+  return new Promise(function(resolve, reject) {
+
+    // Start searching with the preferred port, if specified
+    if (preferredPort !== undefined) {
+      portFinder.basePort = preferredPort;
     }
-    simulator.client = client;
-    deferred.resolve(simulator);
+    
+    portFinder.getPort(function(err, port) {
+      if (err) {
+        reject(err);
+      } else {
+        console.log('got this port', port);
+        resolve(port);
+      }
+    });
   });
-  return deferred.promise;
+
 }
 
-function runB2G(opts) {
-  var commandReady = commandB2G(opts);
-  var portReady = commandReady.then(portIsReady.bind(null, opts.port));
-  return portReady.then(function() {
-    return commandReady;
+
+// Launches the simulator and wait until it's ready to be used
+function launchSimulatorAndWaitUntilReady(options) {
+
+  var port = options.port;
+  var binary = options.binary;
+  var profile = options.profile;
+
+  return new Promise(function(resolve, reject) {
+
+    launchSimulator(options).then(function(simulatorProcess) {
+      waitUntilSimulatorIsReady(port).then(function() {
+        resolve({
+          process: simulatorProcess,
+          pid: simulatorProcess.pid,
+          port: port,
+          binary: binary,
+          profile: profile
+        });
+      }, function(timedOutError) {
+          reject(timedOutError);
+      });
+    }, function(simulatorLaunchError) {
+      reject(simulatorLaunchError);
+    });
+
   });
 }
 
-function findPaths(opts) {
-  return new FXOSSimulators(opts).then(function(result) {
-    if (!result || !result.length) {
-      var errorMsg = 'No simulator found on your machine.';
-      if (opts && opts.version) {
-        errorMsg = 'No simulator for Firefox version ' + opts.version + ' found on your machine.';
+// Launches the simulator in the specified port
+function launchSimulator(options) {
```

Comment:
There are a _lot_ of empty lines in here. I know you like your JS to breathe, but the `} else {` with empty lines around it feels a bit much.